### PR TITLE
Refactor feature_branch post conditions

### PR DESCRIPTION
**Pull request by CassieNi**
Issue Link: https://github.com/wayfair-incubator/pygitops/issues/128

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.12.0] - 2021-08-30
+
+### Changed
+
+* When we exit feature branch context manager, cleaned up the working directory.
+
```
Comment (jdw6359) [inline CHANGELOG.md:14]: ideally, the changes to release the `0.12.0` version of this package will be a separate change

File: tests/util.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+SOME_COMMIT_MESSAGE = "some-commit-message"
+SOME_ACTOR = Actor("some-user", "some-user@wayfair.com")
+SOME_CONTENT = "some-content"
+
+
+@dataclass
+class Repos:
+    local_repo: Repo
+    remote_repo: Repo
```
Comment (jdw6359) [inline tests/util.py:14]: this definition feels very similar to [MultipleTestingRepos](https://github.com/wayfair-incubator/pygitops/blob/main/tests/test_operations.py#L39). is there a way that we can consolidate the definition of `Repos` and `MultipleTestingRepos`?

File: tests/util.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+from dataclasses import dataclass
+from pathlib import Path
+
+from git import Actor, Repo
+
+SOME_COMMIT_MESSAGE = "some-commit-message"
+SOME_ACTOR = Actor("some-user", "some-user@wayfair.com")
```
Comment (jdw6359) [inline tests/util.py:7]: this is a really big nitpick, but I think there are benefits to leaving `wayfair` out of actual code. we currently reference `wayfair` throughout the repo in various configurations and text, but not in code.

feel free to ignore this if you dont see the point!

```suggestion
SOME_ACTOR = Actor("some-user", "some-user@some-company.com")
```

File: tests/util.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    # remote repo must have initial content in order to be cloned
+    new_file_path = remote_repo_path / "foo.txt"
+    new_file_path.write_text("some changes")
+    remote_repo.index.add([str(new_file_path)])
+    remote_repo.index.commit(
+        SOME_COMMIT_MESSAGE, author=SOME_ACTOR, committer=SOME_ACTOR
+    )
+
+    local_repo = Repo.clone_from(remote_repo.working_dir, local_repo_path)
+    return Repos(local_repo=local_repo, remote_repo=remote_repo)
```
Comment (jdw6359) [inline tests/util.py:31]: this function feels very similar to [`initialize_multiple_empty_repos`](https://github.com/wayfair-incubator/pygitops/blob/main/tests/test_operations.py#L698). is there a way that we can consolidate the uses of these functions?

for example, could we have your new test case use the existing `initialize_multiple_empty_repos` function?

I do like the idea of having shared testing module, so maybe a good middle ground would be to extract the current functions and re-use them?

File: tests/util.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def _get_diff(repo: Repo) -> str:
```
Comment (jdw6359) [inline tests/util.py:34]: this function feels very helpful - thank you for writing it!

Comment (jdw6359) [reply tests/util.py:34]: thought: a `diff` is just a general comparison between two states, so it isnt immediately clear that this function will get the difference between the latest commit and and unstaged / untracked changes. 

could we either rename this function to indicate this constraint / intention, or include a comment explaining this?

File: tests/util.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    return repo.git.diff("--cached")
+
+
+def _some_source_modifier(repo: Repo, filename: str) -> None:
+    # Add new file
+    test_file_path = Path(repo.working_dir) / filename
+    test_file_path.touch()
+    # Modify file
+    test_file_path = Path(repo.working_dir) / filename
+    test_file_path.write_text(SOME_CONTENT)
```
Comment (jdw6359) [inline tests/util.py:54]: this function does not feel very reusable, in that it does a very specific thing. do we need a helper function for it?

if there is some source modifier that we want to re-use, perhaps the function definition could be scoped to the individual test case?

File: tests/test_operations.py [hunk 7]
```diff
@@ -68,6 +75,22 @@ def test_stage_commit_push_changes__push_failure__raises_pygitops_error(
             )
 
 
+def test_feature_branch_context_manager__add_new_file_in_different_branches__clean_up_successfully(
```
Comment (jdw6359) [inline tests/test_operations.py:78]: great! I think there are some other scenarios we would want to test as well.

what about changes where:
* an existing file is removed
* an existing file is modified

these test cases will help us ensure that your cleanup implementation does what we would like it to do!

**Review by jdw6359 — COMMENTED** [review]
thank you for taking on this work @CassieNi !

this is super exciting!

this feels like this is on the right track, but needs some more testing for us to be more confident in this behavior. I suspect that there are nuances with the behavior of the `git clean` and `git reset` operations, and we'll want to fully test that before rolling out a new version of the lib

File: tests/util.py (hunk 4 above)
Comment (CassieNi) [reply tests/util.py:31]: Yes! I'll update this.

File: tests/util.py (hunk 6 above)
Comment (CassieNi) [reply tests/util.py:54]: It's been used in two places in one function. I'll rename it to be scoped to the individual test case.

File: tests/test_operations.py (hunk 7 above)
Comment (CassieNi) [reply tests/test_operations.py:78]: make sense! I'll add these test cases.

File: tests/test_operations.py [hunk 8]
```diff
@@ -68,6 +77,103 @@ def test_stage_commit_push_changes__push_failure__raises_pygitops_error(
...
+        ),
+        (
+            _add_new_file,
+            [SOME_OTHER_FILENAME],
+            _delete_existing_file,
+            [SOME_CONTENT_FILENAME],
+            SOME_NEW_FILE_DIFF,
+            SOME_DELETE_FILE_DIFF,
+        ),
+    ],
```
Comment (jdw6359) [inline tests/test_operations.py:153]: this is awesome!

* I love that you are having the source_modifers be callables
* I love that we get to test a variety of these scenarios!

fantastic job designing these tests and making them this reusable 😄

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [0.12.0] - 2021-08-30
+
+### Changed
+
+* Cleaned up the working directory on feature branch when we exit feature branch context manager.
+
```
Comment (jdw6359) [inline CHANGELOG.md:14]:
```suggestion
## [Unreleased]

### Changed

* Cleaned up the working directory on feature branch when we exit feature branch context manager.

```

File: setup.cfg [hunk 10]
```diff
@@ -3,7 +3,7 @@ name = pygitops
 url = https://github.com/wayfair-incubator/pygitops
 author = Josh Woodward
 author_email = josh.woodward2693@gmail.com
-version = 0.11.1
+version = 0.12.0
```
Comment (jdw6359) [inline setup.cfg:6]:
```suggestion
version = 0.11.1
```

**Review by jdw6359 — APPROVED** [review]
lgtm! thanks for doing this!

when you are ready to release version 0.12.0 of this, we'll just want a follow up change with modifications to the `setup.cfg` and the `CHANGELOG.md`

we'll then want to create a release via the github ui, which I can do for you if you are unable.

the package will then be available on pypi!

**Pull request merged**